rm-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/rm/client/monitoring/views/CompactFlowPanelOwn.java

```diff
@@ -0,0 +1,194 @@
+        // own nodes are nodes that already been deployed and used by user
+    }
+
+    @Override
+    public void redrawNodeSource(NodeSource nodeSource) {
+        nodeSourceTiles.get(nodeSource.getSourceName()).refresh(nodeSource);
+        if (isNodeSourceDrawn(nodeSource.getSourceName())) {
+            super.redrawNodeSource(nodeSource);
+        }
+    }
+
+    /**
+     * draws a node and if necessary draws nodesource, and if necessary draws host
+     * @param nodeTile
+     * @param hostTile
+     */
+    @Override
+    protected void drawNormalNode(CompactView.Tile nodeTile, CompactView.Tile hostTile) {
+        int index = 0;
+
+        if (!isNodeSourceDrawn(nodeTile.getNode().getSourceName())) {
+            final CompactView.Tile nsTile = nodeSourceTiles.get(nodeTile.getNode().getSourceName());
+            this.insert(nsTile, size());
+            model.add(new HierarchyNodeSource(nsTile.getNodesource()));
+            model.sort(Comparator.comparing(a -> a.getNodeSource().getSourceName()));
+        }
+
+        for (HierarchyNodeSource hierarchyNodeSource : model) {
+            if (hierarchyNodeSource.getNodeSource().getSourceName().equals(nodeTile.getNode().getSourceName())) {
+                ++index;
+                index += hierarchyNodeSource.getDeploying().size();
+
+                int indexOfFirstHost = index;
+
+                for (HierarchyHost hierarchyHost : hierarchyNodeSource.getHosts()) {
+                    if (hierarchyHost.getHost().getHostName().equals(nodeTile.getNode().getHostName())) {
+                        ++index;
+
+                        hierarchyNodeSource.incrementTiles();
+                        hierarchyHost.incrementTiles();
+
+                        hierarchyHost.getNodes().add(0, nodeTile.getNode());
+
+                        this.insert(nodeTile, index);
+                        return;
+                    } else {
+                        index += hierarchyHost.getTiles();
+                    }
+                }
+
+                // if we are here than there is no host for this node yet
+                // thus we will add it and node after
+                hierarchyNodeSource.incrementTiles();
+                final HierarchyHost hierarchyHost = new HierarchyHost(hostTile.getHost());
+                hierarchyNodeSource.getHosts().add(0, hierarchyHost);
+                this.insert(hostTile, indexOfFirstHost);
+
+                hierarchyNodeSource.incrementTiles();
+                hierarchyHost.incrementTiles();
+                hierarchyHost.getNodes().add(0, nodeTile.getNode());
+                this.insert(nodeTile, indexOfFirstHost + 1);
+
+                return;
+
+            } else {
+                index += hierarchyNodeSource.getTiles();
+            }
+        }
+    }
+
+    /**
+     * removes node tile, and dangling hosts and nodesources
+     * @param node
+     */
+    @Override
+    public void remove(NodeSource.Host.Node node) {
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Refactor this method to reduce its Cognitive Complexity from 38 to the 15 allowed. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3776)